File: rd-kt/rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/Protocol.kt
package com.jetbrains.rd.framework

import com.jetbrains.rd.framework.base.RdExtBase
import com.jetbrains.rd.framework.impl.InternRoot
import com.jetbrains.rd.framework.impl.ProtocolContexts
import com.jetbrains.rd.util.getLogger
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.reactive.ViewableSet
import com.jetbrains.rd.util.string.RName

class Protocol(
    override val name: String,
    override val serializers: ISerializers,
    override val identity: IIdentities,
    override val scheduler: IScheduler,
    override val wire: IWire, //to initialize field with circular dependencies
    val lifetime: Lifetime,
    serializationCtx: SerializationCtx? = null,
    parentContexts: ProtocolContexts? = null
) : IRdDynamic, IProtocol {


    override val location: RName = RName(name)
    override val outOfSyncModels: ViewableSet<RdExtBase> = ViewableSet()

    override val isMaster: Boolean = identity.dynamicKind == IdKind.Client

    companion object {
        val logCategory = "protocol"
        fun sublogger(subcategory: String) = getLogger("$logCategory.$subcategory")
        val initializationLogger = sublogger("INIT")
    }

    override val protocol: IProtocol get() = this
    override val serializationContext: SerializationCtx = serializationCtx ?: SerializationCtx(serializers, mapOf("Protocol" to InternRoot().also {
        it.rdid = RdId.Null.mix("ProtocolInternRoot")
        scheduler.invokeOrQueue {
            it.bind(lifetime, this, "ProtocolInternRoot")
        }
    }))

    override val contexts: ProtocolContexts = parentContexts ?: ProtocolContexts(serializationContext)

    init {
        if (wire is IContextAwareWire)
            wire.contexts = contexts

        if (parentContexts == null) {
            contexts.also {
                it.rdid = RdId.Null.mix("ProtocolContextHandler")
                scheduler.invokeOrQueue {
                    it.bind(lifetime, this, "ProtocolContextHandler")
                }
            }
        }
    }
}

File: rd-cpp/src/rd_framework_cpp/src/main/protocol/Protocol.cpp
#include "protocol/Protocol.h"

#include "serialization/SerializationCtx.h"
#include "intern/InternRoot.h"

#include "spdlog/sinks/stdout_color_sinks.h"

#include <utility>

namespace rd
{
std::shared_ptr<spdlog::logger> Protocol::initializationLogger =
	spdlog::stderr_color_mt<spdlog::synchronous_factory>("initializationLogger", spdlog::color_mode::automatic);

constexpr string_view Protocol::InternRootName;

void Protocol::initialize() const
{
	internRoot = std::make_unique<InternRoot>();

	context = std::make_unique<SerializationCtx>(
		serializers.get(), SerializationCtx::roots_t{{util::getPlatformIndependentHash("Protocol"), internRoot.get()}});

	internRoot->rdid = RdId::Null().mix(InternRootName);
	scheduler->queue([this] { internRoot->bind(lifetime, this, InternRootName); });
}

Protocol::Protocol(std::shared_ptr<Identities> identity, IScheduler* scheduler, std::shared_ptr<IWire> wire, Lifetime lifetime)
	: IProtocol(std::move(identity), scheduler, std::move(wire)), lifetime(lifetime)
{
	// initialize();
}

Protocol::Protocol(Identities::IdKind kind, IScheduler* scheduler, std::shared_ptr<IWire> wire, Lifetime lifetime)
	: IProtocol(std::make_shared<Identities>(kind), scheduler, std::move(wire)), lifetime(lifetime)
{
	// initialize();
}

Protocol::~Protocol() = default;

SerializationCtx& Protocol::get_serialization_context() const
{
	if (!context)
	{
		initialize();
	}
	return *context;
}

}	 // namespace rd
